File: Trees_Graphs/go/FCA.go
package main
import (
  "types"
  "fmt"
)


func getFirstCommonAncestor(node1 *types.TreeNodeWithParent, node2 *types.TreeNodeWithParent) *types.TreeNodeWithParent {
  var path1, path2 types.StringStack
  var parent1, parent2, node *types.TreeNodeWithParent
  var way1, way2 string

  parent1 = node1
  parent2 = node2

  for parent1.Parent != nil {
    if parent1.Parent.LeftChild == parent1 {
      path1.Push("left")
    } else {
      path1.Push("right")
    }

    parent1 = parent1.Parent
  }

  for parent2.Parent != nil {
    if parent2.Parent.LeftChild == parent2 {
      path2.Push("left")
    } else {
      path2.Push("right")
    }

    parent2 = parent2.Parent
  }

  node = parent1

  for {
    way1, _ = path1.Pop()
    way2, _ = path2.Pop()

    if way1 == way2 {
      if way1 == "left" {
        node = node.LeftChild
      } else {
        node = node.RightChild
      }
    } else if node1 == node  || node2 == node {
      return node.Parent
    } else {
      break
    }

  }

  return node
}


func main() {
  root := types.TreeNodeWithParent{Value: 4}

  secondLevelLeftChild := types.TreeNodeWithParent{Value: 2, Parent: &root}
  secondLevelRightChild := types.TreeNodeWithParent{Value: 6, Parent: &root}
  root.LeftChild = &secondLevelLeftChild
  root.RightChild = &secondLevelRightChild

  thirdLevelFirst := types.TreeNodeWithParent{Value: 1, Parent: &secondLevelLeftChild}
  thirdLevelSecond := types.TreeNodeWithParent{Value: 3, Parent: &secondLevelLeftChild}
  thirdLevelThird := types.TreeNodeWithParent{Value: 5, Parent: &secondLevelRightChild}
  thirdLevelFourth := types.TreeNodeWithParent{Value: 7, Parent: &secondLevelRightChild}

  secondLevelLeftChild.LeftChild = &thirdLevelFirst
  secondLevelLeftChild.RightChild = &thirdLevelSecond
  secondLevelRightChild.LeftChild = &thirdLevelThird
  secondLevelRightChild.RightChild = &thirdLevelFourth

  fmt.Println(getFirstCommonAncestor(&thirdLevelThird, &thirdLevelFourth))
}

File: Trees_Graphs/python/FCA.py
from Tree import *
from Stack import Stack


def get_first_common_ancestor(node1, node2):
    print(node1, node2)
    path1, path2 = Stack(), Stack()
    parent1, parent2 = node1, node2

    while parent1.parent is not None:
        if parent1.parent.left_child == parent1:
            path1.push("left")
        else:
            path1.push("right")

        parent1 = parent1.parent

    while parent2.parent is not None:
        if parent2.parent.left_child == parent2:
            path2.push("left")
        else:
            path2.push("right")

        parent2 = parent2.parent

    node = parent1

    while True:
        way1, way2 = path1.pop(), path2.pop()

        if way1 == way2:
            if way1 == "left":
                node = node.left_child
            else:
                node = node.right_child
        elif node1 == node or node2 == node:
            return node.parent
        else:
            break

    return node


def main():
    array = list(range(7))
    tree = Tree(array)
    tree.print_tree()

    print(get_first_common_ancestor(tree.root.left_child.left_child, tree.root.right_child))


if __name__ == '__main__':
    main()
